polygeist/lib/polygeist/Passes/AffineReduction.cpp

```diff
@@ -1,264 +1,367 @@
-    // not used in the current for (see hasAllDimReduced).
-    // The load result are passed to the new forOp as iter args.
-    SmallVector<Value, 4> newIterArgs;
-    llvm::append_range(newIterArgs, forOp.getRegionIterArgs());
-    rewriter.setInsertionPoint(forOp);
-    for (auto pair : candidateOpsInFor) {
-      auto *movedLoad = rewriter.clone(*std::get<0>(pair));
-      newIterArgs.push_back(movedLoad->getResult(0));
+    // Move the load outside the loop (recall that the load is loop invariant).
+    // The load result is passed to the new 'forOp' as iter args.
+    SmallVector<Value, 4> NewIterArgs;
+    llvm::append_range(NewIterArgs, ForOp.getRegionIterArgs());
+    Rewriter.setInsertionPoint(ForOp);
+    for (ReductionOp &Op : ReductionOps) {
+      Operation *MovedLoad = Rewriter.clone(*Op.getLoad());
+      LLVM_DEBUG(llvm::dbgs() << "New Loop:\n" << *MovedLoad << "\n");
+      NewIterArgs.push_back(MovedLoad->getResult(0));
     }
 
-    // create the for.
-    AffineForOp newForOp = rewriter.create<AffineForOp>(
-        forOp.getLoc(), forOp.getLowerBoundOperands(), forOp.getLowerBoundMap(),
-        forOp.getUpperBoundOperands(), forOp.getUpperBoundMap(),
-        forOp.getStep(), newIterArgs);
-
-    // remove load operation inside the for.
-    size_t i = 0;
-    size_t origNumRegionArgs = forOp.getNumRegionIterArgs();
-    for (auto pair : candidateOpsInFor) {
-      std::get<0>(pair)->getResult(0).replaceAllUsesWith(
-          newForOp.getBody()->getArguments()[i + origNumRegionArgs + 1]);
-      rewriter.eraseOp(std::get<0>(pair));
-      ++i;
+    // Create the new loop.
+    AffineForOp NewForOp = Rewriter.create<AffineForOp>(
+        ForOp.getLoc(), ForOp.getLowerBoundOperands(), ForOp.getLowerBoundMap(),
+        ForOp.getUpperBoundOperands(), ForOp.getUpperBoundMap(),
+        ForOp.getStep(), NewIterArgs);
+
+    // Remove the load operation inside the new loop.
+    size_t ArgNo = 0, OrigNumRegionArgs = ForOp.getNumRegionIterArgs();
+    for (ReductionOp &Op : ReductionOps) {
+      Op.getLoad()->getResult(0).replaceAllUsesWith(
+          NewForOp.getBody()->getArguments()[ArgNo + OrigNumRegionArgs + 1]);
+      Rewriter.eraseOp(Op.getLoad());
+      ++ArgNo;
     }
 
-    Block *newBlock = newForOp.getBody();
-    Block *oldBlock = forOp.getBody();
-    SmallVector<Value, 4> newBlockTransferArgs;
-    newBlockTransferArgs.push_back(newForOp.getInductionVar());
-    for (size_t i = 0; i < origNumRegionArgs; i++)
-      newBlockTransferArgs.push_back(newForOp.getRegionIterArgs()[i]);
-    assert(oldBlock->getNumArguments() == newBlockTransferArgs.size() &&
+    Block *NewBlock = NewForOp.getBody(), *OldBlock = ForOp.getBody();
+    SmallVector<Value, 4> NewBlockTransferArgs;
+    NewBlockTransferArgs.push_back(NewForOp.getInductionVar());
+    for (size_t ArgNo = 0; ArgNo < OrigNumRegionArgs; ++ArgNo)
+      NewBlockTransferArgs.push_back(NewForOp.getRegionIterArgs()[ArgNo]);
```

Comment:
Avoid calling `getRegionIterArgs` in the loop every time. Move out

Reply:
sure